### PR TITLE
SearchByCriteria

### PR DESCRIPTION
**Pull request by Yam0r**

### CODE REVIEW

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Screenshot 2024-09-28 at 10 26 06
Branch still has conflicts, fix them first

File: src/main/java/mate/academy/springintro/dto/BookDto.java [hunk 1]
```diff
@@ -13,5 +13,6 @@ public class BookDto {
     private String isbn;
     private BigDecimal price;
     private String description;
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/dto/BookDto.java]: remove empty line

File: src/main/java/mate/academy/springintro/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -21,7 +21,6 @@ public class CreateBookRequestDto {
     @NotNull
     @Min(value = 0)
     private BigDecimal price;
-
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/dto/CreateBookRequestDto.java:24]: Use `@NotBlank` for String

File: src/main/java/mate/academy/springintro/exception/EntityNotFoundException.java [hunk 3]
```diff
@@ -3,7 +3,8 @@
 import java.util.function.Supplier;
 
 public class EntityNotFoundException extends RuntimeException {
-    public EntityNotFoundException(String message) {
+    public EntityNotFoundException(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/exception/EntityNotFoundException.java]:
```suggestion
    public EntityNotFoundException(String message) {
```

File: src/main/java/mate/academy/springintro/repository/SpecificationProvider.java [hunk 4]
```diff
@@ -3,7 +3,8 @@
 import org.springframework.data.jpa.domain.Specification;
 
 public interface SpecificationProvider<T> {
-    Specification<T> getSpecification(String[] params);
+    Specification<T> getSpecification(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/SpecificationProvider.java]:
```suggestion
    Specification<T> getSpecification(String[] params);
```

File: src/main/java/mate/academy/springintro/repository/SpecificationProviderManager.java [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package mate.academy.springintro.repository;
 
 public interface SpecificationProviderManager<T> {
-    SpecificationProvider<T> getSpecificationProvider(String key);
+    SpecificationProvider<T> getSpecificationProvider(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/SpecificationProviderManager.java]:
```suggestion
    SpecificationProvider<T> getSpecificationProvider(String key);
```

File: src/main/java/mate/academy/springintro/repository/book/BookSpecificationBuilder.java [hunk 6]
```diff
@@ -19,7 +19,8 @@ public Specification<Book> build(BookSearchParameters bookSearchParameters) {
         Specification<Book> spec = Specification.where(null);
         if (bookSearchParameters.author() != null && bookSearchParameters.author().length > 0) {
             spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("author")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/book/BookSpecificationBuilder.java]: create constants for String "author" etc

File: src/main/java/mate/academy/springintro/repository/book/spec/AuthorSpecificationProvider.java [hunk 7]
```diff
@@ -10,7 +10,9 @@
 public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
 
     public Specification<Book> getSpecification(String[] params) {
-        return (root, query, criteriaBuilder) -> root.get("author")
+        return (root,
+                query,
+                criteriaBuilder) -> root.get("author")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/book/spec/AuthorSpecificationProvider.java]: create constant for "author"

File: src/main/java/mate/academy/springintro/repository/book/spec/PriceSpecificationProvider.java [hunk 8]
```diff
@@ -10,7 +10,9 @@
 public class PriceSpecificationProvider implements SpecificationProvider<Book> {
 
     public Specification<Book> getSpecification(String[] params) {
-        return (root, query, criteriaBuilder) -> root.get("price")
+        return (root,
+                query,
+                criteriaBuilder) -> root.get("price")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/book/spec/PriceSpecificationProvider.java]: create constant for "price"

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Why it looks like you already merged Search books HW ?

File: src/main/java/mate/academy/springintro/repository/book/BookSpecificationBuilder.java [hunk 9]
```diff
@@ -13,18 +13,25 @@
 public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
 
     private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+    private final String authorMessage = "author";
+    private final String priceMessage = "price";
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/repository/book/BookSpecificationBuilder.java]: Let's use constant formatting 
`private static final String AUTHOR_PARAMETER = "author";`

File: src/main/java/mate/academy/springintro/repository/book/BookSpecificationBuilder.java [hunk 10]
```diff
@@ -13,18 +13,25 @@
...
+        if (bookSearchParameters.price() != null
+                && bookSearchParameters.price().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider(priceMessage)
                     .getSpecification(bookSearchParameters.price()));
         }
         return spec;
     }
 }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/repository/book/BookSpecificationBuilder.java]: Only one empty line at the end of the file ;)

File: src/main/java/mate/academy/springintro/repository/book/spec/AuthorSpecificationProvider.java [hunk 11]
```diff
@@ -8,14 +8,17 @@
 
 @Component
 public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    private final String authorMessage = "author";
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/repository/book/spec/AuthorSpecificationProvider.java]: the same, use constant format

File: src/main/java/mate/academy/springintro/repository/book/spec/PriceSpecificationProvider.java [hunk 12]
```diff
@@ -8,14 +8,17 @@
 
 @Component
 public class PriceSpecificationProvider implements SpecificationProvider<Book> {
+    private final String priceMessage = "price";
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/repository/book/spec/PriceSpecificationProvider.java]: the same

File: src/main/java/mate/academy/springintro/service/BookService.java [hunk 13]
```diff
@@ -12,8 +12,6 @@ public interface BookService {
 
     BookDto findById(Long id);
 
-    void updateBookById(Long id, CreateBookRequestDto updateDto);
-
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/BookService.java:16]: Why do you remove this method?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's fix a few mistakes ;)

File: src/main/java/mate/academy/springintro/repository/book/spec/AuthorSpecificationProvider.java [hunk 14]
```diff
@@ -8,14 +8,17 @@
 
 @Component
 public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String PRICE_PARAMETER = "price";
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/book/spec/AuthorSpecificationProvider.java]: copy-paste, be attentive - `root.get("author")`

File: src/main/java/mate/academy/springintro/controller/BookController.java [hunk 15]
```diff
@@ -42,12 +41,6 @@ public BookDto createBook(@RequestBody @Valid CreateBookRequestDto requestDto) {
         return bookService.save(requestDto);
     }
 
-    @PutMapping("/{id}")
-    public void updateBookById(@PathVariable Long id,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/controller/BookController.java]: why did you delete this endpoint?

File: src/main/java/mate/academy/springintro/repository/book/BookSpecificationProviderManager.java [hunk 16]
```diff
@@ -10,6 +10,7 @@
 @RequiredArgsConstructor
 @Component
 public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/book/BookSpecificationProviderManager.java]: don't add empty line after class / method declaration

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request closed without merging**